### ISSUE TITLE
CDS NOT SENT on sidecars after creating/editing DestinationRule iff subset labels are not matched

### ISSUE THREAD

**Issue by valer-cara** [istio/istio#20120]
**Bug description**
I'm running istio in a multi-cluster with replicated control planes as described in the docs.

When updating a DR, CDS synchronization to sidecars breaks.
Restarting pilot will not fix the issue. Restarting the sidecars seems to re-sync those specific sidecars.

I've built a custom pilot image with more logging for debugging purposes and it seems that for those `NOT_SYNCED` connections, CDSWatch is set to `false`. From my understanding of the flow, i believe that because Envoy does not initiate a CDS watch, pilot will not push any updates to it.

I'm trying to figure out whether that's normal envoy behavior or if it should actually initiate CDS watch when reconnecting to a newly started pilot instance. 

**Expected behavior**
CDS updates are sent to sidecars, `istioctl ps` shows all items as SYNCED

**Steps to reproduce the bug**
**TL;DR** https://github.com/valer-cara/istio-testcase-20120

Hard to reproduce. I've noticed that on apply/delete/edit of the DR in this manifest, the de-sync tends to occur: https://gist.github.com/valer-cara/f69314783f3be5d11dfe3e53ad876e69

After a de-sync, i haven't found a way to make sidecars to resync.

I've noticed that force-restarting the sidecars will get them in a SYNCed state, but if i re-apply or simply edit the destination rule, CDS again goes to STALE.

Edits on the ServiceEntry or VirtualSerice, in the manifest above create no issues and are being propagated. It's only the DR.

**Version (include the output of `istioctl version --remote` and `kubectl version` and `helm version` if you used Helm)**
```
client version: 1.4.0
citadel version: 1.4.3
galley version: 1.4.3
ingressgateway version: 1.4.3
pilot version: unknown
sidecar-injector version: 1.4.3
data plane version: 1.4-alpha.c095f0b1409134b71de04ffb52675e9ce7e5cd88 (4 proxies), 1.4.3 (1 proxies)
```

**How was Istio installed?**
Via the istio-operator, version `1.4-dev`

**Environment where bug was observed (cloud vendor, OS, etc)**
AWS provisioned with https://typhoon.psdn.io/

**Comment by arnecls** [comment]
We have the same issue (#19902).
We are currently looking into locality aware routing as a root cause. We have first seen this appearing after we tested different modes. Also we have one cluster with 1.3, which does not use locality based routing (also because it does not have the common labels required), where this issue is not seen.

**Comment by valer-cara** [comment]
Done some more testing, trying all sorts of "permutation" of the inputs and i think i'm able to reproduce this.

In my scenario, the DR has a few subsets with labels and the ServiceEntry lists endpoints for each of those labels.

If i edit the DR and add a subset with a label that's not currently defined & synced in the cluster; the STALE condition is triggered and the only way to get out of it is to fix the label mismatch in the DR (or remove the label for good) and force restart the sidecars (eg: `pkill -f svc.cluster.local` on the machines -- but *be cautious* not to kill passers by)

**Comment by valer-cara** [comment]
Yup, i can now confirm that the above fixes my issue. :jack_o_lantern: 

The ServiceEntry endpoints matching each of the DR subset labels **must** exist & be synced before creating the DR in order to avoid CDS going `STALE`.

**Comment by valer-cara** [comment]
I've tried to put together a test case for this issue, managed to reproduce it following the steps in the README. https://github.com/valer-cara/istio-testcase-20120

Feedback welcome.

PS: MUTUAL_TLS must be enabled for this to occur

**Comment by arnecls** [comment]
I also have a reproduction rule now, but a different one.

1. Deploy a cluster with:
```
global:
  localityLbSetting:
    enabled: false
```
2. Change the config to
```
global:
  localityLbSetting:
    enabled: true
    distribute:
      - from: "europe-west4/europe-west4-b/*"
        to:
          "europe-west4/europe-west4-b": 90
          "europe-west4/europe-west4-c": 10
      - from: "europe-west4/europe-west4-c/*"
        to:
          "europe-west4/europe-west4-b": 10
          "europe-west4/europe-west4-c": 90
```
And restart pilot.
After the pilot restart is done, _all_ pods are in CDS "NOT SENT" state until the sidecar is restarted and don't accept cluster updates (e.g. outlier detection anymore).
